native/native.tests/tests/org/jetbrains/kotlin/konan/test/blackbox/ProgramNameTest.kt

```diff
@@ -0,0 +1,66 @@
+package org.jetbrains.kotlin.konan.test.blackbox
+
+import org.jetbrains.kotlin.konan.test.blackbox.support.*
+import org.jetbrains.kotlin.konan.test.blackbox.support.compilation.TestCompilationResult.Companion.assertSuccess
+import org.jetbrains.kotlin.konan.test.blackbox.support.util.ClangDistribution
+import org.jetbrains.kotlin.konan.test.blackbox.support.util.compileWithClang
+import org.jetbrains.kotlin.native.executors.runProcess
+import org.junit.jupiter.api.Tag
+import org.junit.jupiter.api.Test
+import java.io.File
+import kotlin.test.assertEquals
+import kotlin.time.Duration.Companion.seconds
+
+@Tag("program-name")
+class ProgramNameTest : AbstractNativeSimpleTest() {
+
+    @Test
+    fun programNameTest() {
+        // 1. Compile main.c to main.cexe
+
+        val cExecutable = buildDir.resolve("main.cexe")
+        compileWithClang(
+            clangDistribution = ClangDistribution.Llvm,
+            sourceFiles = listOf(sourceDir.resolve("main.c")),
+            outputFile = cExecutable,
+            additionalClangFlags = listOf("-Wall", "-Werror"),
+        ).assertSuccess()
+
+        // 2. Compile kotlinPrintEntryPoint.kt to kotlinPrintEntryPoint.kexe
+
+        val kotlinCompilation = compileToExecutableInOneStage(
+            generateTestCaseWithSingleFile(
+                sourceFile = sourceDir.resolve("kotlinPrintEntryPoint.kt"),
+                testKind = TestKind.STANDALONE_NO_TR,
+                extras = TestCase.NoTestRunnerExtras("main")
+            )
+        ).assertSuccess()
+
+        // 3. run main.cexe (with different parameters) to call kotlin executable
+
+        fun validate(expected: String, vararg args: String) {
+            val binaryName = kotlinCompilation.resultingArtifact.executableFile.path
+            val result = runProcess(cExecutable.absolutePath, binaryName, *args) {
+                timeout = 60.seconds
+            }
+            assertEquals("calling exec...\n$expected", result.stdout)
+            assertEquals("", result.stderr)
+        }
+
+        // kotlinPrintEntryPoint removes .kexe
+        validate("programName: app\nargs:", "app.kexe")
+
+        // Simulate a custom program name, see e.g. https://busybox.net/downloads/BusyBox.html#usage
+        validate("programName: customProgramName\nargs:", "customProgramName")
+        validate("programName: customProgramName\nargs: firstArg, secondArg", "customProgramName", "firstArg", "secondArg")
+
+        // No program name - this would not be POSIX compliant, see https://pubs.opengroup.org/onlinepubs/9699919799/functions/exec.html:
+        // "[...] requires a Strictly Conforming POSIX Application to pass at least one argument to the exec function"
+        // However, we should not crash the Kotlin runtime because of this.
+        validate("programName: null\nargs:")
```

Comment:
On Linux, this fails for me with
```text
java.lang.AssertionError: Expected <calling exec...
programName: null
args:>, actual <calling exec...
programName: 
args:>.
```
Please take a look.

Reply:
This is very interesting. Linux behaves here differently as windows & macOS. I fixed the problem, and added more explanation in the code. I used the following C standard PDF: https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf.

Please tell me if you would want to let windows+macOS behave like linux. The way I wrote it now feels more natural to me, however I fully understand that all 3 options have its benefits and drawbacks:

A) The way I wrote it now: all 3 platforms behave the same; no program name leads to `programName=null`, and empty programName leads to `programName=""`.
B) all 3 platforms behave the same; however both no program name and empty program Name lead to `programName=""`. (the linux way of thinking)
C) Fully native behaviour: no programName will be `programName=""` on linux and `programName=null` on macOS/windows; and empty programName will be everywhere `programName=""`.